pkg/fileutils/fileutils_unix.go

```diff
@@ -5,18 +5,51 @@ package fileutils // import "github.com/docker/docker/pkg/fileutils"
 import (
 	"context"
 	"fmt"
+	"io"
 	"os"
 
 	"github.com/containerd/containerd/log"
+	"golang.org/x/sys/unix"
 )
 
 // GetTotalUsedFds Returns the number of used File Descriptors by
 // reading it via /proc filesystem.
 func GetTotalUsedFds() int {
-	if fds, err := os.ReadDir(fmt.Sprintf("/proc/%d/fd", os.Getpid())); err != nil {
-		log.G(context.TODO()).Errorf("Error opening /proc/%d/fd: %s", os.Getpid(), err)
-	} else {
-		return len(fds)
+	name := fmt.Sprintf("/proc/%d/fd", os.Getpid())
+
+	// Fast-path for Linux 6.2 (since [f1f1f2569901ec5b9d425f2e91c09a0e320768f3]).
+	// From the [Linux docs]:
+	//
+	// "The number of open files for the process is stored in 'size' member of
+	// stat() output for /proc/<pid>/fd for fast access."
+	//
+	// [Linux docs]: https://docs.kernel.org/filesystems/proc.html#proc-pid-fd-list-of-symlinks-to-open-files:
+	// [f1f1f2569901ec5b9d425f2e91c09a0e320768f3]: https://github.com/torvalds/linux/commit/f1f1f2569901ec5b9d425f2e91c09a0e320768f3
+	var stat unix.Stat_t
+	if err := unix.Stat(name, &stat); err == nil && stat.Size > 0 {
+		return int(stat.Size)
+	}
+
+	f, err := os.Open(name)
+	if err != nil {
+		log.G(context.TODO()).WithError(err).Error("Error listing file descriptors")
+		return -1
+	}
+	defer f.Close()
+
+	var (
+		fdCount int
+		names   = make([]string, 0, 100)
+	)
+	for {
+		names, err = f.Readdirnames(100)
```

Comment:
Pre-allocating a slice for `names` is a waste as `f.Readdirnames` returns a new slice.
```suggestion
	var fdCount int
	for {
		names, err := f.Readdirnames(100)
```

Reply:
Doh! Good call. I had it because initially I calculated `size(names)` outside the for-loop, then realised that wouldn't work of course 😂